external/java/src/main/java/ch/epfl/dedis/skipchain/SkipchainRPC.java

```diff
@@ -74,39 +84,114 @@ public SkipBlock getSkipblock(SkipblockId id) throws CothorityCommunicationExcep
+            }
 
-        try {
-            SkipchainProto.GetUpdateChainReply reply =
-                    SkipchainProto.GetUpdateChainReply.parseFrom(msg);
-            //TODO: add verification that the skipblock is valid by hashing and comparing to the id
+            SkipBlock start = new SkipBlock(r2.getUpdateList().get(0));
+            if (!trustedLatest.equals(start.getId())) {
+                throw new CothorityCryptoException("first returned block does not match requested hash");
+            }
 
-            if (reply.getUpdateCount() == 0){
-                logger.info("didn't find any updates to {}", scID);
-                return null;
+            // Step through the returned blocks one at a time, verifying
+            // the forward links, and that they link correctly backwards.
+            for (int j = 0; j < r2.getUpdateCount(); j++) {
+                SkipBlock b = new SkipBlock(r2.getUpdateList().get(j));
+                if (j == 0 && update.size() > 0) {
+                    if (Arrays.equals(update.get(update.size()-1).getHash(), b.getHash())) {
+                        continue;
+                    }
+                }
+                if (!b.verifyForwardSignatures()) {
+                    throw new CothorityCryptoException("forward signature verification failed");
+                }
+                // Cannot check back links until we've confirmed the first one
+                if (update.size() > 0) {
+                    if (b.getBackLinks().size() == 0) {
+                        throw new CothorityCryptoException("no backlink");
+                    }
+                    SkipBlock prevBlock = update.get(update.size() - 1);
+                    int link = prevBlock.getHeight();
+                    if (link > b.getHeight()) {
+                        link = b.getHeight();
+                    }
+                    if (!b.getBackLinks().get(link-1).equals(prevBlock.getId())) {
+                        throw new CothorityCryptoException("corresponding backlink doesn't point to previous block");
+                    }
+                    if (!prevBlock.getForwardLinks().get(link-1).getTo().equals(b.getId())) {
+                        throw new CothorityCryptoException("corresponding forwardlink doesn't point to next block");
+                    }
+                }
+                update.add(b);
             }
-            SkipBlock sb = new SkipBlock(reply.getUpdate(reply.getUpdateCount() - 1));
-            logger.info("Got the following latest skipblock: {}", sb);
 
-            return sb;
-        } catch (InvalidProtocolBufferException e) {
-            throw new CothorityCommunicationException(e);
+            SkipBlock last = update.get(update.size()-1);
+
+            // If they updated us to the end of the chain, return.
+            if (last.getForwardLinks().size() == 0) {
+                logger.info("Got the following latest skipblock: {}", Hex.printHexBinary(last.getId().getId()));
```

Comment:
is this left over from debugging?

Reply:
this is actually left from the old code - https://github.com/dedis/cothority/pull/1649/files#diff-e1b7d9d6ed776d2c287512dce30b3a63L105